### PR TITLE
Optimized isEquals() method of RecordFactory implementations

### PR DESCRIPTION
**Pull request by Donnerbart**
After adding some tests I'm quite sure that the `PartitioningStrategy` is not relevant in the `DataRecordFactroy.isEquals()` method. The key is used to calculate the `partitionHash` of a `Data` instance, but that once is not used in the `Data.isEqual()` method. We just compare the payload and its size there.

### CODE REVIEW

**Comment by pveentjer** [comment]
Can you also add a equals check for the same instance? It skips the potentially very expensive deserialization.

```
 @Override
    public boolean isEquals(Object value1, Object value2) {
        if (value1 == null && value2 == null) {
            return true;
        }
        if(value1 == value2){
            return true;
        }
        if (value1 == null) {
            return false;
        }
        if (value2 == null) {
            return false;
        }


        return serializationService.toData(value1).equals(serializationService.toData(value2));
    }
```

**Comment by Donnerbart** [comment]
That's a good one. With all the test cases as security net, I simplified the methods a bit, e.g. the `(value1 == value2)` also covers the `(value1 == null && value2 == null)` case. The `return v1 != null ? v1.equals(v2) : v2 == null` is the IDEA default pattern after the identity and type check.

**Comment by Donnerbart** [comment]
My main question is still, if we need the `PartitionStrategy` in the `isEquals()` method or not. At the moment it doesn't make any different if I add it to the code or not. But I'm also using the default partition strategy. Any ideas of a custom `PartitionStrategy`, that will influence an equals comparison?

**Comment by Donnerbart** [comment]
After adding more tests I'm sure the `PartitioningStrategy` is irrelevant for the `isEquals()` method. This will help us to optimize another PR, since the `isEquals() implementation is just dependent on the in-memory-format.

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java [hunk 1]
```diff
@@ -51,17 +51,11 @@ public void setValue(Record<Object> record, Object value) {
...
             return true;
         }
-        if (v1 == null) {
-            return false;
-        }
-        if (v2 == null) {
-            return false;
-        }
-        return v1.equals(v2);
+        Object v1 = value1 instanceof Data ? serializationService.toObject(value1) : value1;
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java]: Maybe as an optimisation, we could add the 
```
if (value1 == null || value2 == null) {
              return false;
          }
```
before deserialising to skip deserialising if one parameter is `null` but the other one isn't.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java]: Good one! I also added some more test cases with an empty `Data`, so the last comparison still has full coverage (the `toData()` can still return `null` with an empty `Data` instance).

**Comment by Donnerbart** [comment]
Closed, as it's part of https://github.com/hazelcast/hazelcast/pull/11192 now
